### ISSUE TITLE
Are there plans to integrate the new Core Content Blocks Registry / TcaGenerator to set TCA Felds with YAML?

### ISSUE THREAD

**Issue by Kleini13** [b13/container#546]
It wold be a straight forward setup  if the same strategy of config.yaml for creating TCA fields in EXT:Container would be hooked in as with EXT:ContentBlocks. As these two extensions are the replacement for EXT:GridElemenzts for most people. 
Are there any plans from developers in this area? 
https://docs.typo3.org/p/friendsoftypo3/content-blocks/main/en-us/YamlReference/FieldTypes/Index.html#field-types

Advantages:
- Reuse of (Content Blocks) MIxins (Basics)
- No duplicate solutions for field creation when redundant fields are used in EXT:ContentBlocks and EXT:Containers
- Significantly less code
- Use of the new Core Content Block Registry / TcaGenerator (or what ever)

(I am not a TYPO3 expert, so forgive me if I have not expressed myself clearly)

**Comment by nhovratov** [comment]
@Kleini13 You are on the right track. And indeed, this is already possible to a certain extent.

1. Define a Content Block for your container (ContentBlocks/ContentElements/two-column-container/config.yaml):

```
name: vendor/two-column-container
typeName: vendor_two_columns_container
group: container
saveAndClose: true
fields:
  - identifier: header
    useExistingField: true
```

2. Register the Container Configuration (manually) -> Here EXT:container could add a helper method)
Configuration/TCA/Overrides/*.php

```
<?php

use B13\Container\Backend\Preview\ContainerPreviewRenderer;
use B13\Container\Tca\ContainerConfiguration;

$containerConfiguration = new ContainerConfiguration(
    cType: 'vendor_two_columns_container',
    label: '',
    description: '',
    grid: [
        [
            ['name' => 'links', 'colPos' => 200],
            ['name' => 'rechts', 'colPos' => 201],
        ],
    ]
);
$GLOBALS['TCA']['tt_content']['containerConfiguration'][$containerConfiguration->getCType()] = $containerConfiguration->toArray();
// Uncomment for Container < 4.0.0
//$GLOBALS['TCA']['tt_content']['types'][$containerConfiguration->getCType()]['previewRenderer'] = ContainerPreviewRenderer::class;
```

3. Add TypoScript rendering definition

```
tt_content.vendor_two_columns_container {
  dataProcessing {
    100 = B13\Container\DataProcessing\ContainerProcessor
  }
}
```

4. Delete backend-preview.html (only for Container version < 4.0.0)

This way the ContainerPreviewRenderer from EXT:container will kick in and render the grids in the backend.

That's it. From here on you have **full** control over the Container fields. You can even define FlexForm configuration, which is oftentimes needed in YAML (No XML anymore!). Icon (no need to register it anymore) and frontend template are handled by Content Blocks in one place.

TODO: For this to work even easier, everything should be defined at one place, instead of 3. It would make sense to have it all in Content Blocks YAML. This means CB has to define a new type "Container" (special Content Element type), which holds the Container configuration. There is no API for this yet, but it's on my mental list. Could be added as an extension. On the other side, Container has to provide an API to register Container configuration **without** simultanously registering a new Content Element in TCA (see example above).

I hope we can work together to bring these extensions closer together (because they should!). This would even more separate concerns -> Content Blocks for Content Element definitions and Container for Configuration and backend rendering of grids.

EDIT: Since EXT:container version 4.0.0 the preview rendering is no longer done via previewRenderers. Therefore, the TCA override of previewRenderer must be removed (the class no longer exists). Also, you can now provide an own backend-preview for the header and footer section (content section is provided from EXT:container via PSR-14 event).

**Comment by ErHaWeb** [comment]
It's great to see these efforts. I have already successfully combined content blocks and containers in a similar way in a customer project without this description. I would love to be able to control the container behaviour directly in the content block yaml in the future.

**Comment by nhovratov** [comment]
@Kleini13 

> If I can be of any help, tell me ;-)

Of course I am more than happy to get help. For myself I don't know when I will be working on this feature. But in case you want to kickstart it, you can contact me on Slack. I will give you the information you need.

**Comment by Kleini13** [comment]
It works like a charm, but a lot of steps. 
For Everyone who likes to tray:

4. Setup the backend-preview.html (ContentBlocks/ContentElements/two-column-containervendor/templates/backend-preview.html)
 
```
<f:asset.css identifier="content-block-atelier26-atelier26/container-2col-lg-test-be" href="{cb:assetPath()}/preview.css"/>
<html
    xmlns:f="http://typo3.org/ns/TYPO3/CMS/Fluid/ViewHelpers"
    data-namespace-typo3-fluid="true"
>

<f:if condition="{grid}">
    <f:render partial="PageLayout/Grid" arguments="{_all}" />
</f:if>

</html>
```

**Comment by nhovratov** [comment]
It should work without the backend-preview.html. Just delete it and the preview renderer from Container will kick in.

**Comment by Kleini13** [comment]
You are right. It is not taking the backend-preview.html anyways. But what do I have to do if I want to overwrite the b13/container/Resources/Private/Templates/Grid.html with my own backend preview?

**Comment by nhovratov** [comment]
Just had a quick look: This event looks handy: https://github.com/b13/container/blob/b59b76ced4a60c0f93579b6263735ef6d74ccdb7/Classes/Backend/Preview/ContainerPreviewRenderer.php#L112

**Comment by linawolf** [comment]
3,: Mmh so I leave all the typoscript rendering and Templates etc in one place then or also move them?
4.: What backend-preview.html

**Comment by nhovratov** [comment]
> 3,: Mmh so I leave all the typoscript rendering and Templates etc in one place then or also move them? 4.: What backend-preview.html

3: You leave it like it used to be, with the exception that you don't need to specify any template paths anymore, as Content Blocks does this.

4: https://docs.typo3.org/p/friendsoftypo3/content-blocks/main/en-us/Definition/Templates/Index.html#backend-preview-html
